File: swift/MemoryEngine.swift
import Foundation

public struct MemoryNode: Codable {
    var uid: String
    var timestamp: String?
    var path: String?
    var summary: String?
}

private let hybridDir = URL(fileURLWithPath: "memory_vault/hybrid_store")
private let nodesFile = hybridDir.appendingPathComponent("graph_nodes.json")
private let edgesFile = hybridDir.appendingPathComponent("edges_map.json")
private let indexFile = hybridDir.appendingPathComponent("hybrid_index.json")

private func loadJSON<T: Codable>(_ url: URL, defaultValue: T) -> T {
    let fm = FileManager.default
    guard fm.fileExists(atPath: url.path) else { return defaultValue }
    do {
        let data = try Data(contentsOf: url)
        return try JSONDecoder().decode(T.self, from: data)
    } catch {
        return defaultValue
    }
}

private func saveJSON<T: Codable>(_ url: URL, value: T) {
    let fm = FileManager.default
    try? fm.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
    if let data = try? JSONEncoder().encode(value) {
        try? data.write(to: url)
    }
}

public func initHybridStore() {
    let fm = FileManager.default
    try? fm.createDirectory(at: hybridDir, withIntermediateDirectories: true)
    for url in [nodesFile, edgesFile, indexFile] {
        if !fm.fileExists(atPath: url.path) {
            saveJSON(url, value: [String: AnyCodable]())
        }
    }
}

public func addMemoryNode(entry: [String: Any], references: [String]?) -> String? {
    let config: SyntraConfig
    do {
        config = try loadConfig()
    } catch {
        return nil
    }
    guard config.memoryMode == "hybrid" else { return nil }
    initHybridStore()

    let uid = UUID().uuidString
    var nodes: [String: MemoryNode] = loadJSON(nodesFile, defaultValue: [:])
    let node = MemoryNode(
        uid: uid,
        timestamp: entry["timestamp"] as? String,
        path: entry["source"] as? String,
        summary: (entry["summary"] as? String)?.prefix(200).description
    )
    nodes[uid] = node
    saveJSON(nodesFile, value: nodes)

    var edges: [String: [String]] = loadJSON(edgesFile, defaultValue: [:])
    edges[uid] = references ?? []
    saveJSON(edgesFile, value: edges)

    var index: [String: [String]] = loadJSON(indexFile, defaultValue: [:])
    let src = entry["source"] as? String ?? "unknown"
    index[src, default: []].append(uid)
    saveJSON(indexFile, value: index)

    return uid
}

public func traverseMemory() -> [MemoryNode] {
    let config: SyntraConfig
    do {
        config = try loadConfig()
    } catch {
        return []
    }
    var results: [MemoryNode] = []
    if config.memoryMode == "hybrid" {
        initHybridStore()
        let nodes: [String: MemoryNode] = loadJSON(nodesFile, defaultValue: [:])
        results = Array(nodes.values)
    } else {
        let vaults = ["memory_vault/modi", "memory_vault/valon"]
        for v in vaults {
            if let files = try? FileManager.default.contentsOfDirectory(atPath: v) {
                for f in files where f.hasSuffix(".json") {
                    let path = URL(fileURLWithPath: v).appendingPathComponent(f)
                    if let data = try? Data(contentsOf: path),
                       let node = try? JSONDecoder().decode(MemoryNode.self, from: data) {
                        results.append(node)
                    }
                }
            }
        }
    }
    return results
}

struct AnyCodable: Codable {}

File: memory_engine.py
import json
import os
import uuid
from pathlib import Path

from utils.io_tools import load_config

HYBRID_DIR = Path("memory_vault/hybrid_store")
NODES_FILE = HYBRID_DIR / "graph_nodes.json"
EDGES_FILE = HYBRID_DIR / "edges_map.json"
INDEX_FILE = HYBRID_DIR / "hybrid_index.json"


def _load_json(path):
    if path.exists():
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    return {}


def _save_json(path, data):
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2)


def init_hybrid_store():
    """Ensure hybrid storage files exist."""
    HYBRID_DIR.mkdir(parents=True, exist_ok=True)
    for p in (NODES_FILE, EDGES_FILE, INDEX_FILE):
        if not p.exists():
            _save_json(p, {})


def add_memory_node(entry, references=None):
    """Add a memory entry to the hybrid graph and return its UID."""
    config = load_config()
    if config.get("memory_mode", "flat") != "hybrid":
        return None

    init_hybrid_store()
    uid = str(uuid.uuid4())
    nodes = _load_json(NODES_FILE)
    nodes[uid] = {
        "uid": uid,
        "timestamp": entry.get("timestamp"),
        "path": entry.get("source"),
        "summary": entry.get("summary", "")[:200],
    }
    _save_json(NODES_FILE, nodes)

    edges = _load_json(EDGES_FILE)
    edges[uid] = references or []
    _save_json(EDGES_FILE, edges)

    index = _load_json(INDEX_FILE)
    src = entry.get("source") or "unknown"
    index.setdefault(src, []).append(uid)
    _save_json(INDEX_FILE, index)
    return uid


def traverse_memory():
    """Yield memory entries according to the configured mode."""
    config = load_config()
    mode = config.get("memory_mode", "flat")

    if mode == "hybrid":
        init_hybrid_store()
        nodes = _load_json(NODES_FILE)
        for node in nodes.values():
            yield node
    else:
        for vault in [Path("memory_vault/modi"), Path("memory_vault/valon")]:
            if not vault.exists():
                continue
            for file in vault.glob("*.json"):
                with open(file, "r", encoding="utf-8") as f:
                    try:
                        data = json.load(f)
                        yield data
                    except Exception:
                        continue

